DatadogInternal/Sources/NetworkInstrumentation/URLSession/URLSessionInstrumentation.swift

```diff
@@ -0,0 +1,95 @@
+        guard let feature = core.get(feature: NetworkInstrumentationFeature.self) else {
+            throw ProgrammerError(description: "URLSession tracking must be enabled before enabling URLSessionInstrumentation using either RUM or Trace feature.")
+        }
+
+        guard (configuration.delegateClass as AnyObject) is URLSessionTaskDelegate & URLSessionDataDelegate else {
+            throw ProgrammerError(description: "URLSessionInstrumentation requires delegate class to be of type URLSessionTaskDelegate & URLSessionDataDelegate.")
+        }
+
+        try feature.bindIfNeeded(configuration: configuration)
+    }
+
+    /// Disables URLSession instrumentation.
+    /// - Parameters:
+    ///   - delegateClass: The delegate class to unbind.
+    ///   - core: The instance of Datadog SDK to disable URLSession instrumentation in (global instance by default).
+    ///
+    /// Note: delegate type provides compile time safety to ensure correct delegate type is passed.
+    public static func disable(delegateClass: AnyClass, in core: DatadogCoreProtocol = CoreRegistry.default) {
+        do {
+            try disableOrThrow(delegateClass: delegateClass, in: core)
+        } catch let error {
+            consolePrint("\(error)")
+        }
+    }
+
+    internal static func disableOrThrow(delegateClass: AnyClass, in core: DatadogCoreProtocol) throws {
+        guard let feature = core.get(feature: NetworkInstrumentationFeature.self) else {
+            throw ProgrammerError(description: "URLSession tracking must be enabled before enabling URLSessionInstrumentation using either RUM or Trace feature.")
+        }
+
+        guard (delegateClass as AnyObject) is URLSessionTaskDelegate & URLSessionDataDelegate else {
+            throw ProgrammerError(description: "URLSessionInstrumentation requires delegate class to be of type URLSessionTaskDelegate & URLSessionDataDelegate.")
+        }
+
+        feature.unbind(delegateClass: delegateClass)
+    }
+}
+
+extension URLSessionInstrumentation {
+    /// Configuration of URLSession instrumentation.
+    public struct Configuration {
+        /// The delegate class to be used to swizzle URLSessionTaskDelegate & URLSessionDataDelegate methods.
+        public var delegateClass: AnyClass
+
+        /// Additional first party hosts to consider in the interception.
+        public var firstPartyHostsTracing: FirstPartyHostsTracing?
+
+        /// Configuration of URLSession instrumentation.
+        /// - Parameters:
+        ///   - delegate: The delegate class to be used to swizzle URLSessionTaskDelegate & URLSessionDataDelegate methods.
+        ///   - firstPartyHostsTracing: Additional first party hosts to consider in the interception.
+        public init(delegateClass: AnyClass, firstPartyHostsTracing: FirstPartyHostsTracing? = nil) {
+            self.delegateClass = delegateClass
+            self.firstPartyHostsTracing = firstPartyHostsTracing
+        }
+    }
```

Comment:
I think we can constrain the type of the delegate class with:
```swift
/// Configuration of URLSession instrumentation.
public struct Configuration {
    /// The delegate class to be used to swizzle URLSessionTaskDelegate & URLSessionDataDelegate methods.
    public var delegateClass: URLSessionDataDelegate.Type

    /// Additional first party hosts to consider in the interception.
    public var firstPartyHostsTracing: FirstPartyHostsTracing?

    /// Configuration of URLSession instrumentation.
    /// - Parameters:
    ///   - delegate: The delegate class to be used to swizzle URLSessionTaskDelegate & URLSessionDataDelegate methods.
    ///   - firstPartyHostsTracing: Additional first party hosts to consider in the interception.
    public init(delegateClass: URLSessionDataDelegate.Type, firstPartyHostsTracing: FirstPartyHostsTracing? = nil) {
        self.delegateClass = delegateClass
        self.firstPartyHostsTracing = firstPartyHostsTracing
    }
}
```

Reply:
pretty neat - great find.